jupyter-notebooks/scipy-2022-workshop/4_masks_and_filters.ipynb

```diff
@@ -0,0 +1,281 @@
+    "import cv2"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "image_file = \"/content/20170831_172754_101c_3B_AnalyticMS.tif\"\n",
+    "satdat = rasterio.open(image_file)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Allow division by zero\n",
+    "np.seterr(divide='ignore', invalid='ignore')"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "blue, green, red, nir = satdat.read()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Calculate NDWI\n",
+    "ndwi = (green.astype(float) - nir.astype(float)) / (green + nir)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "water_mask = np.full(ndwi.shape, np.nan)\n",
+    "land_mask = np.full(ndwi.shape, np.nan)\n",
+    "\n",
+    "water_mask[ndwi >= 0] = 1\n",
+    "land_mask[ndwi < 0] = 1"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Visualizing Masks"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Let's take a look at our water and land masks:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "# In the below image, purple areas are those identified as water\n",
+    "\n",
+    "plt.figure(1)\n",
+    "plt.imshow(water_mask)\n",
+    "plt.title(\"Houston water mask\")\n",
+    "plt.colorbar()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# In the below image, purple areas are those identified as land\n",
+    "\n",
+    "plt.figure(2)\n",
+    "plt.imshow(land_mask)\n",
+    "plt.title(\"Houston land mask\")\n",
+    "plt.colorbar()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "You might notice that in both images, there are still small portions of water within the land areas, and vice versa. What if we wanted to clean up this mask to get a more general area of where land and water areas are? We can do this by applying morphological filters to our masks."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Apply Filters"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "More info about morphological filters using cv2: https://docs.opencv.org/4.x/d9/d61/tutorial_py_morphological_ops.html"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "closing_kernel_size = 35 #empty pixels within the mask\n",
```

Comment:
Add more context + comment on the line above, to be consistent with our coding standards.
```suggestion
    "# kernel size for empty pixels within the mask\n",
    "closing_kernel_size = 35,"
```